xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletURLFactory.java

```diff
@@ -607,12 +608,60 @@ public URL createAttachmentURL(String filename, String spaces, String name, Stri
                         querystring, xwikidb, context);
                 }
             } catch (XWikiException e) {
-                if (LOGGER.isErrorEnabled()) {
-                    LOGGER.error("Exception while trying to get attachment version !", e);
+                LOGGER.error("Exception while trying to get attachment version !", e);
+            }
+        }
+
+        // In the context of the current doc, no need to reload it.
+        if (isContextDoc(xwikidb, spaces, name, context)) {
+            XWikiAttachment attachment = context.getDoc().getAttachment(filename);
+            if (attachment != null) {
+                return createAttachmentRevisionURL(filename, spaces, name, attachment.getVersion(),
+                    querystring, xwikidb, context);
+            } else {
+                LOGGER.error("Exception while trying to get attachment [{}] from space [{}] and page [{}]!", filename,
+                    spaces, name);
+            }
+        } else {
+            WikiReference originalWikiReference = context.getWikiReference();
+            context.setWikiId(xwikidb);
```

Comment:
This looks dangerous. I don't see where you put back the previous wiki id? i.e. I don't see the try/finally.

Reply:
There's a finally line 640. But yeah it might miss some comments.